### ISSUE TITLE
Prisma migrate uses too many connections

### ISSUE THREAD

**Issue by PanMan** [prisma/prisma#11818]
### Bug description

Hi!
We use prisma migrate on heroku (`npx prisma migrate deploy`). We use a DB instance that has a limit on the number of connections, 10 for the develop environment.
Now often our migrations fail, with an error:
```
`Error querying the database: Server error: `ERROR 42000 (1226): User 'our_username' has exceeded the 'max_user_connections' resource (current value: 10)'
```
When I show the connections (with  `show processlist`) I indeed see the migration using 8 connections (and failing). It does often succeed if I kill all other connections (restart the dyno), and we haven't seen this in production where we have a higher connection limit, but it seems like a bug: I would expect migrate to just use 1 (or 2?) connections, not needing 10.

### How to reproduce

* Limit the amount of connections the DB accepts to eg 8
* See the migration, by `npx prisma migrate deploy` fail

### Expected behavior

Would expect the migration to succeed

### Prisma information

I don't think this is relevant, no queries are executed according to the logs

### Environment & setup

- OS: Linux at Heroku
- Database: MySQL 8.0.23
- Node.js version: v17.4.0

### Prisma Version

```
prisma                : 2.30.3
@prisma/client        : 2.30.3
Current platform      : debian-openssl-1.1.x
Query Engine (Binary) : query-engine b8c35d44de987a9691890b3ddf3e2e7effb9bf20 (at node_modules/@prisma/engines/query-engine-debian-openssl-1.1.x)
Migration Engine      : migration-engine-cli b8c35d44de987a9691890b3ddf3e2e7effb9bf20 (at node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x)
Introspection Engine  : introspection-core b8c35d44de987a9691890b3ddf3e2e7effb9bf20 (at node_modules/@prisma/engines/introspection-engine-debian-openssl-1.1.x)
Format Binary         : prisma-fmt b8c35d44de987a9691890b3ddf3e2e7effb9bf20 (at node_modules/@prisma/engines/prisma-fmt-debian-openssl-1.1.x)
Default Engines Hash  : b8c35d44de987a9691890b3ddf3e2e7effb9bf20
Studio                : 0.423.0
Preview Features      : orderByRelation
```

**Comment by janpio** [comment]
Does the problem go away if you limit the connection pool to 1 via adding `connection_limit=1` to the connection string?

**Comment by PanMan** [comment]
Good one, let me check. However, this is the same connection limit as prisma uses for running, right? Which I want higher than 1... I'll test with 3.

**Comment by janpio** [comment]
Yes, this is really just about pinpointing the problem and understand what is going on. It is obviously not a viable general setting for using Prisma Client and other Prisma tooling.

**Comment by PanMan** [comment]
This does seem to fix it: When deploying with connection limit 3, I never got above 5 connections total (one being myself, for the monitoring). 
Could it be that prisma has a default connection pool size, but the same rules aren't applied to migrate?

**Comment by PanMan** [comment]
btw, I'd like to set the connection limit elsewhere than the URL, as the url (in the heroku case) is "owned" by the addon, and I don't like modifying that (and I probably can't for new review apps(!)). EG an env var or setting in my prisma client settings would work better for me.

**Comment by janpio** [comment]
> This does seem to fix it: When deploying with connection limit 3, I never got above 5 connections total (one being myself, for the monitoring).

Interesting.

> Could it be that prisma has a default connection pool size, but the same rules aren't applied to migrate?

Yes but that should not be used when using `prisma migrate deploy`. Are you also running other commands in the same process?

**Comment by janpio** [comment]
> btw, I'd like to set the connection limit elsewhere than the URL, as the url (in the heroku case) is "owned" by the addon, and I don't like modifying that (and I probably can't for new review apps(!)). EG an env var or setting in my prisma client settings would work better for me.

There are solutions to that already: https://www.prisma.io/docs/guides/development-environment/environment-variables#expanding-variables

But feel free to search for an issue or create your own.

**Comment by PanMan** [comment]
> > Could it be that prisma has a default connection pool size, but the same rules aren't applied to migrate?
> 
> Yes but that should not be used when using `prisma migrate deploy`. Are you also running other commands in the same process?

Nope, we're only doing prisma migrate during migrations (via release phase on Heroku), running `npx prisma migrate deploy` 


> There are solutions to that already: https://www.prisma.io/docs/guides/development-environment/environment-variables#expanding-variables

Aah, perfect, didn't know about that. Thanks! 
Anything else I can test? Problem seems solved for now for me, with the connection limit, but not what I would expect

**Comment by janpio** [comment]
Can you try with a more recent version of Prisma please?
Also, can you possibly share your Prisma schema and/or migration history so we can reproduce exactly what you are running? Might be related to your exact details.

**Comment by janpio** [comment]
Another thought: Are you sure there are not multiple deployments running at the same time, maybe something "preview deployment" or similar?

**Comment by Jolg42** [comment]
Could be something like multiple deployments.

Note that the Migration Engine gets an advisory lock on the database, [see docs](https://www.prisma.io/docs/concepts/components/prisma-migrate#advisory-locking).
Which means that when having multiple deployments at the same time, the first one will get the lock and the following deployments will wait until the lock is released or a 10s timeout throws.
